lib/galaxy/model/security.py

```diff
@@ -1447,62 +1439,171 @@ def get_showable_folders(
 
-    def set_entity_role_associations(self, roles=None, users=None, groups=None, delete_existing_assocs=True):
-        users = users or []
-        roles = roles or []
-        groups = groups or []
-        for role in roles:
-            if delete_existing_assocs:
-                flush_needed = False
-                for a in role.users + role.groups:
-                    self.sa_session.delete(a)
-                    flush_needed = True
-                if flush_needed:
-                    with transaction(self.sa_session):
-                        self.sa_session.commit()
-            for user in users:
-                self.associate_components(user=user, role=role)
-            for group in groups:
-                self.associate_components(group=group, role=role)
+        Associations are set only if a list of new associations is provided.
+        If the provided list is empty, existing associations will be removed.
+        If the provided value is None, existing associations will not be updated.
+        """
+        self._ensure_model_instance_has_id(group)
+        if user_ids is not None:
+            self._set_group_users(group, user_ids)
+        if role_ids is not None:
+            self._set_group_roles(group, role_ids)
+        # Commit only if both group users and group roles have been set.
+        self.sa_session.commit()
+
+    def set_role_user_and_group_associations(
+        self,
+        role: Role,
+        *,
+        user_ids: Optional[List[int]] = None,
+        group_ids: Optional[List[int]] = None,
+    ) -> None:
+        """
+        Set role users and role groups, replacing current associations.
+
+        Associations are set only if a list of new associations is provided.
+        If the provided list is empty, existing associations will be removed.
+        If the provided value is None, existing associations will not be updated.
+        """
+        self._ensure_model_instance_has_id(role)
+        if user_ids is not None:
+            self._set_role_users(role, user_ids or [])
+        if group_ids is not None:
+            self._set_role_groups(role, group_ids or [])
+        # Commit only if both role users and role groups have been set.
+        self.sa_session.commit()
+
+    def _set_user_groups(self, user, group_ids):
+        delete_stmt = delete(UserGroupAssociation).where(UserGroupAssociation.user_id == user.id)
+        insert_values = [{"user_id": user.id, "group_id": group_id} for group_id in group_ids]
+        self._set_associations(user, UserGroupAssociation, delete_stmt, insert_values)
+
+    def _set_user_roles(self, user, role_ids):
+        # Do not include user's private role association in delete statement.
+        delete_stmt = delete(UserRoleAssociation).where(UserRoleAssociation.user_id == user.id)
+        private_role = get_private_user_role(user, self.sa_session)
+        if not private_role:
+            log.warning("User %s does not have a private role assigned", user)
```

Comment:
This likely won't be seen by anyone. If we do need this we should raise an exception here.

Reply:
I agree that an exception is more appropriate. However, it appears there's something going on in user private role processing: we either don't assign a private role in some edge cases, or we delete them. As a result, we have 551 users without a private role on main:
```
galaxy_main=> select count(*) from galaxy_user u where u.id not in (select ura.user_id from user_role_association ura inner join role r on ura.role_id = r.id and r.type = 'private');
 count 
-------
   511
```

I'll post the probable cause on the committers channel. This PR solves it (if that indeed is the cause), but since there are records of users w/o a private role in the database, I think we should fix those records before raising an error here.